statsmodels/tools/tools.py

```diff
@@ -136,15 +137,47 @@ def categorical(data, col=None, dictnames=False, drop=False, ):
     >>> design2 = sm.tools.categorical(struct_ar, col='str_instr', drop=True)
     '''
     if isinstance(col, (list, tuple)):
-        try:
-            assert len(col) == 1
+        if len(col) == 1:
             col = col[0]
-        except:
+        else:
             raise ValueError("Can only convert one column at a time")
+    if not isinstance(col, (string_types, int)) and col is not None:
+        raise TypeError('col must be a str, int or None')
 
     # TODO: add a NameValidator function
     # catch recarrays and structured arrays
-    if data.dtype.names or data.__class__ is np.recarray:
+    if isinstance(data, (pd.Series, pd.DataFrame)):
+        if isinstance(data, pd.DataFrame):
+            if col is None:
+                raise TypeError('col must be a str or int when using a '
+                                'DataFrame')
+            else:
+                if col in data:
+                    data = data[col]
```

Comment:
I think this may cause a problem if the dataframe has integer column names and col is a nonzero int.

Possibly easier to follow if dataframe and series had separate code paths.

sm.categorical should be deprecated entirely.  dictnames even more so, since it is only used in a helper function in discrete_model.

Reply:
It is like ix.  It tries loc and then falls back to iloc for integers.  This is why ix went away.

Reply:
I claim the following will incorrectly raise on L168:
```
arr = np.random.randn(3, 2)
data = pd.DataFrame(arr)
categorical(data, col=1)
```

Handling loc/iloc ambiguity is not something we should allow to become our problem.  Since DataFrame inputs currently raise, make them _always_ raise and tell the user to pass a single Series instead.

Reply:
(or raise on Series too and tell users to use pd.get_dummies directly)

Note: in a recent branch I tried to have sm.categorical wrap pd.Categorical, but some of the recarray test cases broke in exciting ways.

Reply:
I require column names now.